File: leetcode/0064/0064.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
      int row = grid.size();
      int col = grid[0].size();
      vector<vector<int>> dp(row, vector<int>(col, 0));
      for(int i = 0; i < row; i++) {
        for(int j = 0; j < col; j++) {
          if(i == 0 && j == 0) {
            dp[i][j] = grid[i][j];
          } else if(i == 0) {
            dp[i][j] = dp[i][j-1] + grid[i][j];
          } else if(j == 0) {
            dp[i][j] = dp[i-1][j] + grid[i][j];
          } else {
            dp[i][j] = min(dp[i-1][j],dp[i][j-1]) + grid[i][j];
          }
        } 
      }
      return dp[row-1][col-1];
    }
};

int main() {
  vector<vector<int>> grid1 = {{1, 3, 1},{1, 5, 1},{4, 2, 1}};
  vector<vector<int>> grid2 = {{0}};
  cout << Solution().minPathSum(grid1) << endl;
  cout << Solution().minPathSum(grid2) << endl;
  system("PAUSE");
  return 0;
}

File: leetcode/0064/0064.js
/**
 * WARNING 递归解法超时
 * @param {number[][]} grid
 * @return {number}
 */
function getSum(arr, dp, i, j) {
  if(i == 0 || j == 0) {
    return dp[i][j]
  } else  {
    return arr[i][j] + Math.min(getSum(arr, dp, i-1, j), getSum(arr, dp, i, j-1))
  }
}

var minPathSum = function(arr) {
  const row = arr.length
  const col = arr[0].length
  const dp = new Array(row).fill(0).map(() => new Array(col).fill(0))
  arr[0].forEach((e, i) => {
    if(i >= 1) {
      dp[0][i] = dp[0][i-1] + e
    } else {
      dp[0][i] = e
    }
  })
  arr.forEach((e, i) => {
    if(i >= 1) {
      dp[i][0] = dp[i-1][0] + e[0]
    } else {
      dp[i][0] = e[0]
    }
  })
  const sum = getSum(arr, dp, row - 1, col - 1);
  return sum
}